### PR TITLE
Rewrote head

### PR DESCRIPTION
**Pull request by Mikadore**
The head utility is in a pretty poor state. This rewrite is to address that.

## Newly supported features:
- moved to clap
- support all CLI arguments + NUM syntax (`-c 5b`, `-1c`,  `-n 4kiB`) (better output for `--help` too).
- Some bug fixes & feature support, e.g. the following tests fail on the current master:
    - test_head::test_byte_syntax
    - test_head::test_line_syntax
    - test_head::test_negative_zero_lines
    - test_head::test_zero_terminated_syntax_2
- Support for arbitrary data (the current version relies on utf8 encoding when it filters by lines)

We're close to 100% compatibility, albeit I messed up tracking all the features, so there are some obscurities & edge cases.
The most notable of these is that head will write a newline if not present as the last character of output ***if in terminal***.
It won't do this if the output is piped, or redirected to a file. There is a wrapper for `isatty` on crates.io (+ the windows counterpart), which could be helpful.

## On performance & memory:
The performance of `head` can certainly be improved.
Parsing argument syntax is the most obvious here, I will also do some more detailed benchmarking.
Here's a small table of my local measurements (using hyperfine for perf and `valgrind` for memory (heap)):

### Noop:
 |`head -c 0` | Time | Memory |
 |--|--|--|
GNU | 0.5 ms | 44 bytes
Current| 0.8 ms | 25,575 bytes
Rewrite|1.6 ms | 489,631 bytes

#### Bytes (on data from /dev/random):
First 512 bytes of a random file (size 1k)
|`head -c 512 {random_file}`| Time | Memory
|--|--|--|
GNU| 0.5 ms |  1k
Current| 1.5 ms | 17k
Rewrite| 1.6 ms | 534k

First 32 *mega* bytes of a random file (size ~513M)
|`head -c 33554432 {random_file} > /dev/zero`|Time|Memory
|--|--|--|
GNU|7.5 ms|4k
Current|3074 ms| 17k
Rewrite|6.3 ms|534k

*All but the last* 32 mega bytes of a random file (size ~513M)
|`head -c -33554432 {random_file} > /dev/zero`|Time|Memory
|--|--|--|
GNU|99.5 ms|4k
Current| FAILED| FAILED
Rewrite|63.2 ms|534k

Note: I couldn't get the current version to work, the error was
```
head: error: Argument to option 'c' missing
```

#### Lines (on The complete works of Shakespeare):
The whole file is ~5.4M

First 2048 lines:
|`head -n 2048 {text} > /dev/zero`|Time|Memory
|--|--|--|
GNU| 0.6ms | 4k
Current| 1.8ms | 100k
Rewrite| 1.5ms | 534k

All ***but*** the last 4096 lines
|`head -n -4096 {text} > /dev/zero`|Time|Memory
|--|--|--|
GNU|1.9ms|4k
Current|52.5ms|5942k
Rewrite|2.3ms|534k

**Note: this benchmark only includes data read from files, which the rewrite optimizes for. This also means that memory usage isn't really put to test, better benchmarks *can* be made**

In summary, the rewrite has some constant startup and memory issues but is superior to both GNU and the current version in terms of scalability, and medium+ sized workloads, for run time, and superior to just the current version in terms of memory.
I think the startup time could be improved and will be doing some work in this area in the following days/weeks.

## Tests:
I un-ignored most tests and added one. I think testing can & should be improved.
Some tests were not correct (i.e. they'd assert output not compatible with GNU), and were fixed.

### CODE REVIEW

File: src/uu/head/src/constants.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+
+pub const EXIT_FAILURE: i32 = 1;
```
Comment (sylvestre) [inline src/uu/head/src/constants.rs:2]: I value a lot consistency as we have a lot of binaries.
Please move that into head.rs and follow what is done elsewhere.
example:
https://github.com/uutils/coreutils/blob/master/src/uu/ls/src/ls.rs#L43-L52
and
https://github.com/uutils/coreutils/blob/master/src/uu/ls/src/ls.rs#L78-L82

or
https://github.com/uutils/coreutils/blob/master/src/uu/chroot/src/chroot.rs#L21-L40

thanks and sorry for the extra work

File: tests/by-util/test_head.rs [hunk 2]
```diff
@@ -179,3 +158,23 @@ fn test_no_such_file_or_directory() {
...
+        .pipe_in_fixture("sequence")
+        .run()
+        .stdout_is_fixture("sequence.expected");
+}
+
+#[test]
+fn test_zero_terminated() {
+    new_ucmd!()
+        .args(&["-z", "zero_terminated.txt"])
+        .run()
```
Comment (sylvestre) [inline tests/by-util/test_head.rs:178]: please run rustfmt :)

File: src/uu/head/src/app.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
+use crate::constants;
```
Comment (sylvestre) [inline src/uu/head/src/app.rs:1]: please rename this file to head.rs to match the other programs

File: src/uu/head/src/app.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+use crate::parse;
+use clap::{App, Arg};
+use std::ffi::OsString;
+use uucore::{executable, exit};
+
+pub fn app<'a>() -> App<'a, 'a> {
+    App::new(executable!())
```
Comment (sylvestre) [inline src/uu/head/src/app.rs:8]: please follow the other programs example (see below)

**Review by sylvestre — CHANGES_REQUESTED** [review]
Looks great. Just some requests to be consistent with the rest of the programs

File: src/uu/head/src/app.rs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+                                        self.mode = Some(Modes::Bytes(n));
+                                        Some(OsString::from("-c"))
+                                    } else {
+                                        self.mode = Some(Modes::Lines(n));
+                                        Some(OsString::from("-n"))
+                                    }
+                                }
+                                Err(e) => match e {
+                                    parse::ParseError::Overflow => {
+                                        eprintln!(
```
Comment (ycd) [inline src/uu/head/src/app.rs:127]: crash! macro used widely for this operation in codebase, it 'd be better to use it for consistency.

File: src/uu/head/src/app.rs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+                                            "head: Value too large for defined datatype: '{}'",
+                                            arg_str
+                                        );
+                                        exit!(constants::EXIT_FAILURE);
+                                    }
+                                    parse::ParseError::Syntax => {
+                                        eprintln!("head: bad number: '{}'", arg_str);
```
Comment (ycd) [inline src/uu/head/src/app.rs:134]: same here

**Review by ycd — COMMENTED** [review]
Looks great, just a small nitpick, if it is working can you add a test for https://github.com/uutils/coreutils/issues/1799?

**Comment by ycd** [comment]
@Mikadore benchmarks looks interesting and the difference in memory usage looks huge,  which version of GNU coreutils did you used?

File: src/uu/head/src/constants.rs (hunk 1 above)
Comment (Mikadore) [reply src/uu/head/src/constants.rs:2]: Do I understand you correctly that I should move the content of `app.rs` to `head.rs`?

**Comment by Mikadore** [comment]
@ycd `head (GNU coreutils) 8.32`

The memory usage would be more interesting for piped data, where it is necessary to buffer *some of it*, imo.
Tbh, it's just late here and I didn't have the energy to run more of these. 

Overall, I think benchmarking this project as a whole (and comparing it against GNU) would be interesting to see, maybe I'll get to it some time.

**Comment by Mikadore** [comment]
> Looks great, just a small nitpick, if it is working can you add a test for #1799?

I can't add a test for this as the API can't handle invalid strings for stdout.

File: src/uu/head/src/head.rs [hunk 7]
```diff
@@ -1,240 +1,300 @@
...
+    let mut stdout = stdout.lock();
 
-impl Default for Settings {
-    fn default() -> Settings {
-        Settings {
-            mode: FilterMode::Lines(10),
-            verbose: false,
-            zero_terminated: false,
+    loop {
+        let read = input.read(&mut readbuf)?;
```
Comment (Arcterus) [inline src/uu/head/src/head.rs:22]: This (and basically everywhere else using `read()`) should check for `ErrorKind::Interrupted`.

File: src/uu/head/src/parse.rs [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+                match &unit.as_str().as_bytes()[0] {
+                    b'k' | b'K' => {
+                        if is_base_2 {
+                            1024u128.pow(1)
+                        } else {
+                            1000u128.pow(1)
+                        }
+                    }
+                    b'm' | b'M' => {
+                        if is_base_2 {
```
Comment (Arcterus) [inline src/uu/head/src/parse.rs:88]: Since all of these branches are the same, restructure it like so:
```rust
let exp = match blah {
    b'k' | b'K' => 1,
    b'm' | b'M' => 2,
    // and so on...
};

if is_base_2 {
    1024u128.pow(exp)
} else {
    1000u128.pow(exp)
}
```

I think there's also `.to_ascii_lowercase()` or something function you can use to avoid checking both the lowercase and uppercase versions of the letters.

Comment (Arcterus) [inline src/uu/head/src/parse.rs:88]: Since all of these branches are the same, restructure it like so:
```rust
let exp = match blah {
    b'k' | b'K' => 1,
    b'm' | b'M' => 2,
    // and so on...
};

if is_base_2 {
    1024u128.pow(exp)
} else {
    1000u128.pow(exp)
}
```

I think there's also `.to_ascii_lowercase()` or something function you can use to avoid checking both the lowercase and uppercase versions of the letters.

File: src/uu/head/src/parse.rs [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    if src.is_empty() {
+        return Err(ParseError::Syntax);
+    }
+
+    // (1) is the number
+    // (2) is b
+    // (3) is any of the other units
+    // (4) is `i`
+    // (5) is `B`
+    let re = regex::Regex::new(r"^-?(\d+)?(?:(b)|([kKmMgGtTpPeEzZyY])(?:(i)B|(B))?)?$").unwrap();
```
Comment (Arcterus) [inline src/uu/head/src/parse.rs:59]: I would personally prefer this not use a regex, as I think they're hard to read.  Same for `parse_obsolete()`.

File: src/uu/head/src/app.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+                    Some(oss) => {
+                        if oss == "head" {
+                            return Some(oss);
+                        }
+                        let arg_str = &oss.clone().into_string().unwrap_or_else(|_| "".to_owned());
+                        if let Some(res) = parse::parse_obsolete(&arg_str) {
+                            match res {
+                                Ok((n, b)) => {
+                                    if b {
+                                        self.mode = Some(Modes::Bytes(n));
```
Comment (Arcterus) [inline src/uu/head/src/app.rs:118]: This code is pretty deeply nested, so it'd be nice if it was split into functions.

**Review by Arcterus — CHANGES_REQUESTED** [review]
I'm reviewing this on my phone, so I'll see if I can read the code more in-depth later.

**Comment by ycd** [comment]
> @ycd head (GNU coreutils) 8.32

Ah i see, as far as i remember old head (GNU coreutils <= 8.21) had a similar memory usage because it was allocating the memory up front based on the value passed to `-c`.

File: src/uu/head/src/constants.rs (hunk 1 above)
Comment (ycd) [reply src/uu/head/src/constants.rs:1]: Move constants to `app.rs` and instead of functions use an options mod. 

Instead of this:

```rust
#[inline(always)]
pub fn bytes_name() -> &'static str {
    "BYTES"
}

#[inline(always)]
pub fn lines_name() -> &'static str {
    "LINES"
}
```


Use this:

```rust
pub mod Options {
    pub static LINES: &str = "lines";
    pub static BYTES: &str = "bytes";
}
```

Instead of this:

```rust
#[inline(always)]
pub fn usage() -> &'static str {
    "head [FLAG]... [FILE]..."
}
```

Use this:

```rust
static USAGE: &str = "head [FLAG]... [FILE]...";
```

etc.

File: src/uu/head/src/parse.rs (hunk 9 above)
Comment (Mikadore) [reply src/uu/head/src/parse.rs:59]: > I would personally prefer this not use a regex, as I think they're hard to read. Same for `parse_obsolete()`.

I was planning to rewrite this anyway for performance, however, I'm not sure about the readability aspect of it

**Comment by Mikadore** [comment]
I just realized that head supports arguments in the form of `head -123zv`, this means that `parse_obsolete` needs some adjustments in the API + new tests

**Comment by Mikadore** [comment]
I'm happy to report a significant speedup with the rewritten version of the argument parsing. From my basic measurements, basic argument parsing is about 30-60% faster.

We also now completely support the obsolete syntax (e.g. `-123zvqvcvmk` (- this is bad, please don't do this)).

I am however unhappy with the implementation side of things (it's ugly), and the fact that we have some (big? - need to profile) overhead for supporting the obsolete syntax. The only way for this to be resolved, really, (imo) is to either
- stop using clap 
- get clap to support -123abc

The former is very unappealing in regards to consistency and simplicity, and the latter might not be wanted in clap.

However, it's also worth mentioning that our performance is really good, that is, ~0.5ms slower than GNU for very small workloads, and even faster beyond that! Of course, the faster the better, but I think if left as is even the current performance is satisfactory.

File: src/uu/head/src/parse.rs [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+        );
+        assert_eq!(
+            obsolete("-10000000000000000000000"),
+            Some(Err(ParseError::Overflow))
+        );
+    }
+    #[test]
+    #[cfg(target_pointer_width = "32")]
+    fn test_parse_obsolete_overflow_x32() {
+        assert_eq!(
```
Comment (sylvestre) [inline src/uu/head/src/parse.rs:270]: could you please fix the errors here?

File: src/uu/head/src/head.rs [hunk 12]
```diff
@@ -1,240 +1,556 @@
...
+use std::ffi::OsString;
+use std::io::{ErrorKind, Read, Seek, SeekFrom, Write};
+use uucore::{crash, executable, show_error};
+
+const EXIT_FAILURE: i32 = 1;
+const EXIT_SUCCESS: i32 = 0;
+const BUF_SIZE: usize = 65536;
+
+mod options {
+    pub const VERSION: &str = env!("CARGO_PKG_VERSION");
```
Comment (sylvestre) [inline src/uu/head/src/head.rs:12]: please move usage, about and version out of "mod options" (like in the other programs)

**Review by sylvestre — COMMENTED** [review]
cool stuff :)
bravo

**Comment by ycd** [comment]
> get clap to support -123abc

Can you elaborate this? which cases are failing without `parse_obsolete`? 

We had a similar case in `echo`, i fixed it by allowing the leading hyphens (see: https://github.com/uutils/coreutils/pull/1887)

**Comment by Mikadore** [comment]
> > get clap to support -123abc
> 
> Can you elaborate this? which cases are failing without `parse_obsolete`?
> 
> We had a similar case in `echo`, i fixed it by allowing the leading hyphens (see: #1887)

This isn't the same (as far as I see), in  `head -123zvc`, the meaning of `zvc` changes than if it were `head -123 -zvc` (let alone that `-zvc` needs an argument for the `c` part).
 `head -123zvc` is parsed as (in order) "print the first 123" (default is lines), "use \0 as a line terminator", "set verbose mode", "print the first 123 bytes". The `c` isn't the same as `--bytes` here.

So, clap would need some sort of NUM type of argument, in the form of `-NUM{modifiers}`, it would then also need to parse flags correctly, and sort out ambiguities (NUM `c` vs normal `c`).

If there is something like this in clap, I'd be more than happy to use it, but I don't know of it.

**Comment by sylvestre** [comment]
it is ready to be reviewed?
btw, bravo for 
```
 codecov/project Successful in 1s — 63.48% (+0.59%) compared to 52997b6 
```

**Comment by Mikadore** [comment]
> it is ready to be reviewed?
> btw, bravo for
> 
> ```
>  codecov/project Successful in 1s — 63.48% (+0.59%) compared to 52997b6 
> ```

Thanks!

And yes, I think it's ready.

File: src/uu/head/src/head.rs [hunk 13]
```diff
@@ -1,240 +1,557 @@
...
+    \n\
+    With no FILE, or when FILE is -, read standard input.\n\
+    \n\
+    Mandatory arguments to long flags are mandatory for short flags too.\
+    ";
+const USAGE: &str = "head [FLAG]... [FILE]...";
+
+mod options {
+    pub const BYTES_NAME: &str = "BYTES";
+    pub const BYTES_HELP: &str = "\
```
Comment (sylvestre) [inline src/uu/head/src/head.rs:24]: for consistency, please move that in the clap call

File: src/uu/head/src/head.rs [hunk 14]
```diff
@@ -1,240 +1,557 @@
...
+        print the first NUM bytes of each file;\n\
+        with the leading '-', print all but the last\n\
+        NUM bytes of each file\
+        ";
+    pub const LINES_NAME: &str = "LINES";
+    pub const LINES_HELP: &str = "\
```
Comment (sylvestre) [inline src/uu/head/src/head.rs:30]: same

File: src/uu/head/src/head.rs [hunk 15]
```diff
@@ -1,240 +1,557 @@
...
+    pub const VERBOSE_NAME: &str = "VERBOSE";
+    pub const VERBOSE_HELP: &str = "always print headers giving file names";
+    pub const ZERO_NAME: &str = "ZERO";
+    pub const ZERO_HELP: &str = "line delimiter is NUL, not newline";
+    pub const FILES_NAME: &str = "FILE";
+}
+mod parse;
+mod split;
+
+fn app<'a>() -> App<'a, 'a> {
```
Comment (sylvestre) [inline src/uu/head/src/head.rs:46]: please keep the same profile as others
https://github.com/uutils/coreutils/blob/master/src/uu/ls/src/ls.rs#L359
pub fn uumain(args: impl uucore::Args) -> i32 {

File: src/uu/head/src/head.rs [hunk 16]
```diff
@@ -1,240 +1,557 @@
...
-    mode: FilterMode,
-    verbose: bool,
-    zero_terminated: bool,
+pub fn parse_mode<F>(src: &str, closure: F) -> Result<(Modes, bool), String>
+where
+    F: FnOnce(usize) -> Modes,
+{
+    match parse::parse_num(src) {
+        Ok((n, last)) => Ok((closure(n), last)),
+        Err(reason) => match reason {
```
Comment (sylvestre) [inline src/uu/head/src/head.rs:107]: would it be possible to add a test covering this case?

File: src/uu/head/src/head.rs [hunk 17]
```diff
@@ -1,240 +1,557 @@
...
+fn arg_iterate<'a>(
+    mut args: impl uucore::Args + 'a,
+) -> Result<Box<dyn Iterator<Item = OsString> + 'a>, String> {
+    // argv[0] is always present
+    let first = args.next().unwrap();
+    if let Some(second) = args.next() {
+        if let Some(s) = second.to_str() {
+            match parse::parse_obsolete(s) {
+                Some(Ok(iter)) => Ok(Box::new(vec![first].into_iter().chain(iter).chain(args))),
+                Some(Err(e)) => match e {
```
Comment (sylvestre) [inline src/uu/head/src/head.rs:125]: ditto

File: src/uu/head/src/head.rs [hunk 18]
```diff
@@ -1,240 +1,557 @@
...
+    })
 }
 
-// TODO: handle errors on read
-fn head<T: Read>(reader: &mut BufReader<T>, settings: &Settings) -> bool {
-    match settings.mode {
-        FilterMode::Bytes(count) => {
-            for byte in reader.bytes().take(count) {
-                print!("{}", byte.unwrap() as char);
+fn head_backwards_file(input: &mut std::fs::File, options: &HeadOptions) -> std::io::Result<()> {
```
Comment (sylvestre) [inline src/uu/head/src/head.rs:347]: Looking at the number of warnings, it doesn't seem tested ?

**Review by sylvestre — CHANGES_REQUESTED** [review]
Nothing critical but a few requests :)

File: src/uu/head/src/head.rs (hunk 15 above)
Comment (Mikadore) [reply src/uu/head/src/head.rs:46]: I disagree with this change.

having an `app` function that constructs the clap `App` means we can unit test argument parsing. 
The alternative would be to have 2 copies of the same code, one for the unit tests, the other for the actual executable, which would be a nightmare to maintain, but also reduce readability.

File: src/uu/head/src/head.rs (hunk 18 above)
Comment (Mikadore) [reply src/uu/head/src/head.rs:347]: Oh, yeah. Looking at the tests, all inputs are piped.

File: src/uu/head/src/head.rs (hunk 15 above)
Comment (sylvestre) [reply src/uu/head/src/head.rs:46]: ok, no worries

File: src/uu/head/src/head.rs (hunk 18 above)
Comment (Mikadore) [reply src/uu/head/src/head.rs:347]: > Looking at the number of warnings, it doesn't seem tested ?

btw, where do you see these warnings?

Comment (sylvestre) [reply src/uu/head/src/head.rs:347]: Here: https://github.com/uutils/coreutils/pull/1911/files 
when the coverage run are completed

You can also what it here:
https://codecov.io/gh/uutils/coreutils/commit/4ee61a9e033609b16e233e6e61ce34ab63b4ee58/

Comment (sylvestre) [reply src/uu/head/src/head.rs:347]: and here: https://codecov.io/gh/uutils/coreutils/src/3b8dcbf0cfe99657ce7e9b59c8d4f73db33ff5b6/src/uu/head/src/head.rs

File: src/uu/head/src/head.rs (hunk 17 above)
Comment (Mikadore) [reply src/uu/head/src/head.rs:124]: done

File: src/uu/head/src/head.rs (hunk 16 above)
Comment (Mikadore) [reply src/uu/head/src/head.rs:106]: done

File: src/uu/head/src/head.rs (hunk 13 above)
Comment (Mikadore) [reply src/uu/head/src/head.rs]: Done (I'm assuming you're referring only to the `HELP` messages)

File: src/uu/head/src/head.rs (hunk 18 above)
Comment (Mikadore) [reply src/uu/head/src/head.rs:346]: oh nice, didn't know that

**Comment by sylvestre** [comment]
Bravo for this impressive work.

**Comment by Mikadore** [comment]
> Bravo for this impressive work.

Thank you very much! This was my first contribution to a real open source project, and it's been a blast!

Looking forward to work on other things as well
